### PR TITLE
Run code inside an IPython kernel and capture artifacts

### PR DESCRIPTION
**Pull request by StefanoFioravanzo**
This PR introduces the following changes:

1. Step's code is handles as a list of code blocks, to keep information about the original notebook code cells
2. Utils to create an ipython kernel on the fly and run some code in it, while monitoring for stdout, stderr and error and redirecting the to the main thread.
3. Utils to parse the rich outputs produced by (2) and parse them to create a single html artifact for each execution
4. Extension to the code generation module to generate a pipeline that takes advantage of (1) (2) and (3)

### CODE REVIEW

**Comment by StefanoFioravanzo** [comment]
@elikatsis I updated the PR with a better handling of the watcher thread. The relevant changes as

- The watcher thread is created with `daemon=True`, so that it will make the process hand when the main thread exists
- The main thread kills the ipython kernel when the code execution is finished, so that resources are disposed of gracefully (sockets etc..)

In this way we can call `kc.iopub_channel.get_msg` without a timeout and not get into trouble with long running codes.

File: setup.py [hunk 1]
```diff
@@ -33,7 +33,11 @@
         'jinja2 >=2.10, <3.0',
         'graphviz >=0.13, <1.0',
         'pyflakes >=2.1.1',
-        'dill >=0.3, <0.4'
+        'dill >=0.3, <0.4',
+        'IPython >= 7.6.0',
+        'jupyter-client >= 5.3.4',
```
Comment (elikatsis) [inline setup.py]: + jupyter-core >= 4.6.0
+ kfp-server-api == 0.1.18.3

Comment (elikatsis) [reply setup.py]: We can also go for
+ kfp >= 0.2

Comment (elikatsis) [reply setup.py]: + `packaging` is a pytest requirement, which we only install on `dev` installation, but you're using it on `jupyter_utils`, so you should also add an entry for that here.

**Comment by elikatsis** [comment]
I also got this on final auto snapshot step:
```
This step is in Error state with this message: failed to save outputs: path /final_auto_snapshot.html does not exist (or /final_auto_snapshot.html is empty) in archive /argo/outputs/artifacts/final_auto_snapshot.tgz
```

This means that you are adding a `/final_auto_snapshot.html` output artifact, but you are not generating it.
So you should add an if statement to not add the following for the `final_auto_snapshot` step (in `pipeline_template.jinja2` file):
```
mlpipeline_ui_metadata = {'{{ name }}': '/{{ name }}.html'}
```

These will be simplified in the future, when we declare component outputs via function outputs (`->` notation).

**Pull request merged**